### PR TITLE
[0.4.4] Add new features

### PR DESCRIPTION
**Pull request by canattofilipe**
Add the following new features:
- Firmware update automatically (dojot/dojot#1502)
- Publish device status (dojot/dojot#1557)
- Discard black listed values (dojot/dojot#1704 )

### CODE REVIEW

**Comment by mprevide** [comment]
The first 5 commits should probably not be here.

**Comment by canattofilipe** [comment]
> The first 4 commits should probably not be here.
> @canattofilipe

You are right, these commits isn't part of the features added, I will remove it from this PR.

File: src/main/java/org/cpqd/iotagent/ImageDownloader.java [hunk 1]
```diff
@@ -45,12 +47,13 @@ public ImageDownloader(String imageManagerUri, String dataDir,
...
 
-    public String downloadImageAndGenerateUri(String tenant, String imageLabel,
-        String version, int protocol) {
+	public String downloadImageAndGenerateUri(String tenant, String imageLabel, String version, int protocol,
+			String imageId, Map<String, String> queryParams) {
 
         String imageFilename = null;
         try {
-            imageFilename = this.fetchImage(tenant, imageLabel, version);
+			imageFilename = imageId == null ? this.fetchImage(tenant, imageLabel, version)
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/ImageDownloader.java:55]: Fix the indentation

File: src/main/java/org/cpqd/iotagent/ImageDownloader.java [hunk 2]
```diff
@@ -79,6 +82,13 @@ public String downloadImageAndGenerateUri(String tenant, String imageLabel,
                 this.mLogger.error("Unknown protocol: " + protocol);
                 throw new RuntimeException("Failed to generate URI");
         }
+		
+		if (queryParams != null) {
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/ImageDownloader.java:86]: Fix the indentation

File: src/main/java/org/cpqd/iotagent/LwM2MAgent.java [hunk 3]
```diff
@@ -138,7 +143,8 @@ private void sendsUriToDevice(Registration registration, String imageLabel,
 
             String fileUri = null;
             try {
-                fileUri = imageDownloader.downloadImageAndGenerateUri(tenant, imageLabel, newFwVersion, supportedProtocol);
+				fileUri = imageDownloader.downloadImageAndGenerateUri(tenant, imageLabel, newFwVersion,
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/LwM2MAgent.java:146]: Fix the indentation

**Review by mprevide — APPROVED** [review]

**Review by thiagodpf — APPROVED** [review]
The code makes sense to me.
As there is no _lint_ enabled to point out the lines that are indented with tabs, I believe that the PR is approved.

**Pull request merged**